### PR TITLE
Add Ultimate HQ subsidiary block

### PR DESCRIPTION
**Pull request by paulgain**
## Sandbox changes
https://github.com/uktrade/data-hub-sandbox/commit/db7b24165f03df2e6e992303696d61f66f2e9e18

## Add Ultimate HQ subsidiary block 
"Data Hub contains `n` other company record related to this company"

This is hidden behind a feature flag `companies-ultimate-hq`

## Test instructions
1. add an Ultimate HQ company such as `Alphabet Inc` via the new D&B `Add a company` journey.
2. add a subsidiary company such as `Nest Labs, Inc` as above.
3. Navigate to the `Alphabet Inc` companies page.
 
## Screenshots
add-subsidiaries-to-ultimate-hq

## Checklist

[//]: # "When submitting a PR make sure the code review guidelines have been satisfied. 
https://github.com/uktrade/data-hub-frontend/blob/develop/docs/Code%20review%20guidelines.md"

- [ ] Has the branch been rebased to develop?
- [ ] Automated tests (Any of the following when applicable: Unit, Functional or Acceptance)
- [ ] Manual compatibility testing (Browsers: Chrome, Firefox, IE11, Safari)

### CODE REVIEW

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js [hunk 1]
```diff
@@ -64,12 +63,23 @@ describe('Activity feed controllers', () => {
   })
 
   describe('#renderActivityFeed', () => {
+    beforeEach(() => {
+      this.getUltimateHQSubsidiariesStub = sinon.stub().resolves({ results: [{}, {}] })
```
Comment (ian-leggett) [inline src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: We need to remove all occurrences of the `this` in arrow functions.

File: src/apps/companies/apps/activity-feed/controllers.js [hunk 2]
```diff
@@ -1,22 +1,30 @@
...
+  const { token } = req.session
   try {
     const addContentProps = company.archived ? {} : {
       addContentText: 'Add interaction',
       addContentLink: `/companies/${company.id}/interactions/create`,
     }
 
+    const props = {
+      ...addContentProps,
+      apiEndpoint: `/companies/${company.id}/activity/data`,
```
Comment (rafenden) [inline src/apps/companies/apps/activity-feed/controllers.js]: We should use the `urls` module.

File: src/apps/companies/views/template.njk [hunk 3]
```diff
@@ -45,13 +46,19 @@
...
+    {% if isUltimate or isOneListTier %}
       <div class="c-local-header__description">
-        <p>This is an account managed company (One List {{ company.one_list_group_tier.name }})</p>
-        <p>
-          Global Account Manager: {{ company.one_list_group_global_account_manager.name }}
-          <a href="/companies/{{ company.id }}/advisers">View core team</a>
-        </p>
+        {% if isUltimate %}
+          {% set link =  props.subsidiaryCount + ' other company ' + 'record' | pluralise(props.subsidiaryCount) %}
+          <p>Data Hub contains <a class="subsidiaries" href="/companies/{{ company.id }}/subsidiaries">{{ link }}</a> related to this company</p>
```
Comment (web-bert) [inline src/apps/companies/views/template.njk]: Add to the `urls` module and use here

File: src/apps/companies/views/template.njk [hunk 4]
```diff
@@ -45,13 +46,19 @@
...
+        {% endif %}
+        {% if isOneListTier %}
+          <p>This is an account managed company (One List {{ company.one_list_group_tier.name }})</p>
+          <p>
+            Global Account Manager: {{ company.one_list_group_global_account_manager.name }}
+            <a href="/companies/{{ company.id }}/advisers">View core team</a>
```
Comment (web-bert) [inline src/apps/companies/views/template.njk]: Add to the `urls` module and use here

File: test/functional/cypress/specs/companies/company-global-ultimate-spec.js [hunk 5]
```diff
@@ -7,8 +7,13 @@ describe('Company Global Ultimate HQ', () => {
       cy.visit(`/companies/${fixtures.company.dnbGlobalUltimate.id}/activity`)
```
Comment (web-bert) [inline test/functional/cypress/specs/companies/company-global-ultimate-spec.js]: Add to the `urls` module and use here

File: src/apps/companies/apps/activity-feed/controllers.js [hunk 6]
```diff
@@ -1,22 +1,30 @@
 const { fetchActivityFeed } = require('./repos')
+const { getUltimateHQSubsidiaries } = require('../../repos')
 
 async function renderActivityFeed (req, res, next) {
   const { company } = res.locals
+  const { token } = req.session
   try {
     const addContentProps = company.archived ? {} : {
       addContentText: 'Add interaction',
       addContentLink: `/companies/${company.id}/interactions/create`,
```
Comment (web-bert) [inline src/apps/companies/apps/activity-feed/controllers.js]: Add to the `urls` module and use here

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js [hunk 7]
```diff
@@ -124,6 +135,7 @@ describe('Activity feed controllers', () => {
         expect(this.middlewareParameters.resMock.render).to.be.calledOnceWithExactly('companies/apps/activity-feed/views/client-container', {
           props: {
             apiEndpoint: '/companies/dcdabbc9-1781-e411-8955-e4115bead28a/activity/data',
```
Comment (web-bert) [inline src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: Use `urls` module

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js [hunk 8]
```diff
@@ -87,6 +97,7 @@ describe('Activity feed controllers', () => {
               addContentLink: '/companies/dcdabbc9-1781-e411-8955-e4115bead28a/interactions/create',
               addContentText: 'Add interaction',
               apiEndpoint: '/companies/dcdabbc9-1781-e411-8955-e4115bead28a/activity/data',
```
Comment (web-bert) [inline src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: Use `urls` module

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js [hunk 9]
```diff
@@ -87,6 +97,7 @@ describe('Activity feed controllers', () => {
               addContentLink: '/companies/dcdabbc9-1781-e411-8955-e4115bead28a/interactions/create',
```
Comment (web-bert) [inline src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: Use `urls` module

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js [hunk 10]
```diff
@@ -1,5 +1,3 @@
-const { renderActivityFeed } = require('../controllers')
-
 const buildMiddlewareParameters = require('~/test/unit/helpers/middleware-parameters-builder')
 const activityFeedEsFixtures = require('~/test/unit/data/activity-feed/activity-feed-from-es')
 const companyMock = require('~/test/unit/data/company.json')
```
Comment (web-bert) [inline src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: Remove `wavy`

Comment (paulgain) [reply src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: :thumbsdown:
```
const buildMiddlewareParameters = require('../../../../../../test/unit/helpers/middleware-parameters-builder')
const activityFeedEsFixtures = require('../../../../../../test/unit/data/activity-feed/activity-feed-from-es.json')
const companyMock = require('../../../../../../test/unit/data/company.json')
```
@web-bert I don't think we should be rolling this out, we need a better solution as @rafenden has previosuly mentioned. If we continue to roll this out we are creating tech debt.

Comment (web-bert) [reply src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: I think the solution to this problem is to inject these into the tests as helpers, I did this on market access...

## Mocks:
https://github.com/uktrade/market-access-frontend/blob/develop/src/test/app/helpers/mocks.js

### Usage:
https://github.com/uktrade/market-access-frontend/blob/develop/src/app/middleware/dashboard-data.spec.js#L18

## Get data:
https://github.com/uktrade/market-access-frontend/blob/develop/src/test/app/helpers/get-fake-data.js

### Usage:
https://github.com/uktrade/market-access-frontend/blob/develop/src/app/controllers/find-a-barrier.spec.js#L68

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js (hunk 1 above)
Comment (paulgain) [reply src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: I don't think we should be taking `this` out of every single unit test we touch. It's going to take too long to change everything within a file. A better approach is... if an existing file uses `this` just follow suite, if it's a new file then do not introduce `this`, this approach maintains consistency. I'm sorry but I have more important things to do than refactor unit tests where the value of the outcome is zero.

File: src/apps/companies/apps/activity-feed/__test__/controllers.test.js (hunk 10 above)
Comment (paulgain) [reply src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: I don't think we need that, all we need to do is to resolve the paths from root, we don't need to touch test code. 
Personally I prefer the `buildMiddlewareParameters` option.

There are two approaches and probably more:
1. Rafs idea
2.  Set `NODE_PATH` to `.` and define the full path (like Java package imports)

Comment (web-bert) [reply src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: But the paths are only long because we are trying to access items in the root `test` folder, so why not add some helpers to access those things for us without importing them directly, it would make the code more DRY and avoid the long paths.

We already have it setup: https://github.com/uktrade/data-hub-frontend/blob/develop/test/unit/common.js

Comment (paulgain) [reply src/apps/companies/apps/activity-feed/__test__/controllers.test.js]: There's no need for helpers if it's done correctly.

File: src/apps/companies/apps/activity-feed/controllers.js [hunk 11]
```diff
@@ -16,20 +19,26 @@ async function renderActivityFeed (req, res, next) {
...
+    }
+
+    if (company.is_global_ultimate) {
+      const subsidiaries = await getUltimateHQSubsidiaries(token, company.global_ultimate_duns_number)
+      // Substract the Utlimate HQ from the count
+      props.subsidiaryCount = subsidiaries.results.length - 1
+    }
+
     res
       .breadcrumb(company.name, `/companies/${company.id}`)
```
Comment (web-bert) [inline src/apps/companies/apps/activity-feed/controllers.js]: `urls` can be used here
